File: hashmap/contains-duplicate-II/contains-duplicate-II.go
func containsNearbyDuplicate(nums []int, k int) bool {
    last := make(map[int]int, len(nums))

	for i,v := range nums {
		if prev, ok := last[v]; ok && i-prev <= k {
			return true
		}
		last[v] = i
	}

	return false
}

func containsNearbyDuplicateWindow(nums []int, k int) bool {
    window := make(map[int]struct{})

	for i,v := range nums {
		if i > k {
			delete(window, nums[i-k-1])
		}
		if _, ok := window[v]; ok {
			return true
		}
		window[v] = struct{}{}
	}

	return false
}

File: hashmap/contains-duplicate-II/contains-duplicate-II.ts
function containsNearbyDuplicate(nums: number[], k: number): boolean {
    const lastIdx = new Map<number, number>();

    for (let i = 0; i < nums.length; i++){
        const v = nums[i];
        if (lastIdx.has(v)){
            const prev = lastIdx.get(v)!;
            if (i - prev <= k) return true;
        }
        lastIdx.set(v, i);
    }

    return false;
};

function containsNearbyDuplicateWindow(nums: number[], k: number): boolean {
    const window = new Set<number>();

    for (let i = 0; i < nums.length; i++){
        if (i > k) window.delete(nums[i - k - 1]);
        if (window.has(nums[i])) return true;
        window.add(nums[i]);
    }

    return false;
};